pkg/rdt/rdt.go

```diff
@@ -0,0 +1,90 @@
+// +build linux
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package rdt
+
+import (
+	"fmt"
+
+	"github.com/containerd/containerd/log"
+	"github.com/intel/goresctrl/pkg/rdt"
+	"github.com/pkg/errors"
+
+	"github.com/containerd/containerd/plugin"
+)
+
+const (
+	// ResctrlPrefix is the prefix used for class/closid directories under the resctrl filesystem
+	ResctrlPrefix = ""
+)
+
+var enabled bool
+
+// Config contains the configuration of the RDT plugin
+type Config struct {
+	ConfigFile string `toml:"config_file" json:"configFile"`
+}
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		Type:   plugin.InternalPlugin,
+		ID:     "rdt",
+		Config: &Config{},
+		InitFn: initRdt,
+	})
+}
+
+func initRdt(ic *plugin.InitContext) (interface{}, error) {
+	enabled = false
+
+	if err := rdt.Initialize(ResctrlPrefix); err != nil {
+		log.L.Infof("RDT is not enabled: %v", err)
+		return nil, nil
+	}
+
+	config, ok := ic.Config.(*Config)
+	if !ok {
+		return nil, errors.New("invalid RDT configuration")
+	}
+
+	if config.ConfigFile == "" {
+		log.L.Info("No RDT config file specified, RDT not configured")
+		return nil, nil
+	}
+
+	if err := rdt.SetConfigFromFile(config.ConfigFile, true); err != nil {
+		return nil, errors.Wrap(err, "configuring RDT failed")
+	}
+
+	enabled = true
+
+	return nil, nil
```

Comment:
I'm not sure if we should use plugins this way.
Usually when we do
```golang
i, err := p.Instance()
if err != nil {
    return nil, err
}
```
we expect `i` to be a valid object that is safe to use.

Would `sync.Once` be a better fit here?

Reply:
So we try to initialize on first invoke of `ContainerClassFromAnnotations`

Reply:
Either `sync.Once` or `sync.RWMutex` around `enabled` seems like a better approach.

Reply:
> I'm not sure if we should use plugins this way.
> Usually when we do
> 
> ```go
> i, err := p.Instance()
> if err != nil {
>     return nil, err
> }
> ```
> 
> we expect `i` to be a valid object that is safe to use.
> 
> Would `sync.Once` be a better fit here?

The reason I implemented this as a plugin was that it provided a nice integration point, both in terms of configuration and initialization at runtime startup time.

What I would like to have is to bail out at (runtime) startup phase in case of invalid configuration, for example. And not suddenly error out at the first occasion somebody tries to assign a container to RDT. How is the object returned by InitFn supposed to be used? Is there a certain interface that `InitFn` of internal plugins is expected to implement or some other way to achieve this? How about returning an empty struct?

Any advice or thoughts?

Reply:
Any comments @mxpv @samuelkarp ?

Reply:
Woof @mxpv @samuelkarp! Any thoughts on that^? The corresponding PR in CRI-O (cri-o/cri-o#4830) was merged a while ago so I'd like to get containerd on par with that 😉

Reply:
since this seems to be a cri issue on the containerd server, feel free to add your init to the cri init.. we can always move this out of the cri plugin and into it's own plugin down the road if need

on the run once thing something like:
```
var rdtEnabled bool
var rdtEnabledOnce sync.Once

func GetRdtEnabled() {
   return rdtEnabled
}

then in the init.. 
    var result Error
    rdtEnabledOnce.Do(func() {
	rdtEnabled = true // if passes all the enabled checks
        // should be false and set the result error that cause the fail..
	})
       return (some go func for monitoring this capability?), result 
```

If you don't move your init to the cri plugin, and a monitoring thread is not needed for this plugin, I guess it's ok to just stub it? @dmcgowan thoughts?

Reply:
Well, it's not CRI only. It is possible to set the RDT class with `ctr`, too. E.g. something like
```
ctr run --rdt-class my-class ...
```